File: compiler/bench/floyd.fs
\ expected: 90500
\ Floyd-Warshall all-pairs shortest path, 100 nodes

100 constant N
1000000 constant INF
create dist N N * cells allot

: idx ( i j -- addr ) swap N * + cells dist + ;

: init ( -- )
  N 0 do N 0 do
    i j = if 0 else INF then i j idx !
  loop loop
  N 1 do
    1 i i 1- idx !
    1 i 1- i idx !
  loop
  N 10 do
    2 i i 10 - idx !
    2 i 10 - i idx !
  loop ;

: min ( a b -- min ) 2dup > if swap then drop ;

: floyd ( -- )
  N 0 do
    N 0 do
      N 0 do
        j k idx @
        j i idx @ i k idx @ + min
        j k idx !
      loop
    loop
  loop ;

: sum-dist ( -- n )
  0 N 0 do N 0 do
    j i idx @ dup INF < if + else drop then
  loop loop ;

: main init floyd sum-dist . cr ;

File: compiler/bench/floyd.c
// expected: 90500
#include <stdio.h>
#define N 100
#define INF 1000000
long dist[N][N];

void init() {
    for (int i = 0; i < N; i++)
        for (int j = 0; j < N; j++)
            dist[i][j] = (i == j) ? 0 : INF;
    for (int i = 1; i < N; i++) {
        dist[i][i-1] = 1;
        dist[i-1][i] = 1;
    }
    for (int i = 10; i < N; i++) {
        dist[i][i-10] = 2;
        dist[i-10][i] = 2;
    }
}

void floyd() {
    for (int k = 0; k < N; k++)
        for (int i = 0; i < N; i++)
            for (int j = 0; j < N; j++) {
                long alt = dist[i][k] + dist[k][j];
                if (alt < dist[i][j])
                    dist[i][j] = alt;
            }
}

int main() {
    init();
    floyd();
    long sum = 0;
    for (int i = 0; i < N; i++)
        for (int j = 0; j < N; j++)
            if (dist[i][j] < INF)
                sum += dist[i][j];
    printf("%ld\n", sum);
    return 0;
}
